vaadin-flow-components-shared-parent/vaadin-flow-components-base/src/main/java/com/vaadin/flow/component/shared/internal/DisableOnClickController.java

```diff
@@ -0,0 +1,132 @@
+ * to prevent multiple clicks or submissions while the server processes the
+ * event.
+ * <p>
+ * This controller requires that the component implements both
+ * {@link HasEnabled} and {@link ClickNotifier}.
+ *
+ * @param <C>
+ *            Type of the component that uses this controller.
+ */
+@JsModule("./disableOnClickFunctions.js")
+public class DisableOnClickController<C extends Component & HasEnabled & ClickNotifier<C>>
+        implements Serializable {
+
+    private final C component;
+    private boolean disableOnClick = false;
+    private boolean initializedInCurrentRoundTrip;
+
+    /**
+     * Creates a new controller for the given component.
+     *
+     * @param component
+     *            the component to control, not {@code null}
+     */
+    public DisableOnClickController(C component) {
+        this.component = Objects.requireNonNull(component);
+        component.addClickListener(event -> {
+            if (isDisableOnClick()) {
+                component.setEnabled(false);
+            }
+        });
+    }
+
+    /**
+     * Sets whether the component should be disabled when clicked.
+     * <p>
+     * When set to {@code true}, the component will be immediately disabled on
+     * the client-side when clicked, preventing further clicks until re-enabled
+     * from the server-side.
+     *
+     * @param disableOnClick
+     *            whether the component should be disabled when clicked
+     */
+    public void setDisableOnClick(boolean disableOnClick) {
+        this.disableOnClick = disableOnClick;
+        if (disableOnClick) {
+            component.getElement().setProperty("disableOnClick", "true");
+            initDisableOnClick();
+        } else {
+            component.getElement().removeProperty("disableOnClick");
+        }
+    }
+
+    /**
+     * Gets whether the component is set to be disabled when clicked.
+     *
+     * @return whether the component is set to be disabled on click
+     */
+    public boolean isDisableOnClick() {
+        return disableOnClick;
+    }
+
+    /**
+     * Forces the client-side component's {@code disabled} property to be
+     * updated immediately.
+     * <p>
+     * This method should be called from the component's
+     * {@link HasEnabled#setEnabled} method.
+     *
+     * @param enabled
+     *            value to set
+     */
+    public void onSetEnabled(boolean enabled) {
+        // If the component is then disabled and re-enabled during the same
+        // round trip, Flow will not detect any changes and the client side
+        // component would not be enabled again.
+        component.getElement().executeJs("this.disabled = $0", !enabled);
+    }
+
+    /**
+     * Initialize client side disabling so disabled if immediate on click even
+     * if server-side handling takes some time.
+     */
+    private void initDisableOnClick() {
+        if (initializedInCurrentRoundTrip) {
```

Comment:
This name is potentially confusing, as this doesn't reset if the component is already attached. It also looks like it's now possible to add multiple attach listeners.

Maybe always early return this method if there is already an attach listener, because then it never needs to run again.

Reply:
Updated the names and refactored it to return early.

Reply:
Pushed the changes.